Fix bug with low sum of neutrino masses

c-d-leonard opened the pull request:
There was a bug which meant that when defining a cosmology within the normal or inverted hierarchy convention for neutrinos and with a mass below the respective physical limit in each case (~0.059 and ~0.1) you would fall into an infinite loop in ccl_parameters_create when trying to split the sum of masses according to the hierarchy instead of exiting gracefully with an error.

This PR fixes the bug and avoids the infinite loop but doesn't manage to pass an informative error to python because the problem arises in ccl_parameters_create, where there is not yet a cosmology object and therefore no way to pass the error message back to python. I'm leaving this for later (maybe if we deprecate the ccl_parameters_create function in C).

See Issue #552 for discussion.

beckermr commented on src/ccl_core.c:
```diff
@@ -424,19 +425,26 @@ ccl_parameters ccl_parameters_create(
+    //ccl_raise_exception(*status, "The neutrino mass is too low to satisfy \
+    the mass difference measurements in the case of normal hierarchy.\
+    In this low mass case, set mnu_type to sum_inverted or sum_equal.");
+    //  }
```
Let's remove this extra code.

c-d-leonard commented:
Thanks, good call, I'll make those changes.

beckermr commented on pyccl/core.py:
```diff
@@ -579,10 +579,22 @@ def _build_parameters(
+                  and (m_nu[0]>1e-15)):
+                raise ValueError("if mnu_type= sum_inverted, we are using the "
+					                 "inverted hierarchy and so m_nu must "
+					                 "be less than (~)0.0978")    			
```
Oy. flake8 is going to be mean to you. :(

c-d-leonard replied:
I did go run flake8 locally immediately after pushing that commit :P it should be okay now.

c-d-leonard commented:
The current status is: I added a check from the python that the sum of neutrino masses is above the minimum allowed value in the given hierarchy, so we now get an informative error in python if that is done wrong. I also added a test in python to make sure that error gets raised.

I changed the C so that if you call ccl_parameters_create directly from C and you make the same kind of too-low choice for a sum of mnu given your hiearchy, we exit with an informative error. I manually checked that this behaves correctly if I do this, but I'm not sure if there's an elegant way to add a test in C to the suite because of the lack of cosmo status message (there's no cosmology object yet when we're still setting up the parameters object, so we can't pass it back out). What we want the C to do is just exit gracefully, which it is doing - but I don't know how to test it's doing that without just stopping the test suite.

beckermr commented:
Right. At this point we either:

1. Significantly change the C API.
2. Deprecate the C API so that we are sure the test from Python is sufficient.
3. Figure out some other magic that IDK.

My motivation to deprecate the C API is because of shit like this that is super annoying and time consuming.

beckermr commented:
OTOH, the python part is already done! It is a sign from the heavens!

damonge commented:
@beckermr can we forget about the unit test and revisit once we have deprecated the C API?
Since this is a more general problem (testing errors while creating parameters), should we open an issue for that and move on?

beckermr commented:
Seems fine to me.
